src/engine/threaded_engine_perdevice.cc

```diff
@@ -50,6 +50,37 @@ class ThreadedEnginePerDevice : public ThreadedEngine {
   static auto constexpr kWorkerQueue = kFIFO;
 
   ThreadedEnginePerDevice() noexcept(false) {
+    this->Start();
+#ifndef _WIN32
+    pthread_atfork(
+      []() {
+        Engine::Get()->WaitForAll();
+        Engine::Get()->Stop();
+      },
+      []() {
+        Engine::Get()->Start();
+      },
+      []() {
+        Engine::Get()->Start();
+      });
+#endif
+  }
+  ~ThreadedEnginePerDevice() noexcept(false) {
+    gpu_normal_workers_.SignalForKill();
+    gpu_copy_workers_.SignalForKill();
+    cpu_normal_workers_.SignalForKill();
+    this->Stop();
+  }
+
+  void Stop() override {
+    SignalQueuesForKill();
+    gpu_normal_workers_.Clear();
+    gpu_copy_workers_.Clear();
```

Comment:
In LazyAllocArray, I suggest changing exit_now_ name to is_clearing_ or something like that, and then setting it back to false when the Clear() completes.
Before doing that, you probably can clear head_ and more_ , or at least set them to think they're empty and to start from the beginning again when filling.